test/State.hs

```diff
@@ -67,3 +76,28 @@ test m a s (Run runState) =
   , testProperty "put updates the state variable" . forall (s :. s :. m a :. Nil) $
     \ s s' m -> runState (s, put s' >> m) === runState (s', m)
   ]
+
+problematic :: Applicative f => f a
+problematic = (pure (error "should immediately")) <*> pure (error "terminate")
+
+lazy
+  :: (Monad m, Functor f)
+  => GenTerm (f ())
+  -> Run f g m
+  -> [TestTree]
+lazy s (Run run) =
+  [ testProperty "fmap is non-strict" . forall (a :. s :. Nil) $
+    \ a s -> void . eval . run $ (const a <$> pure (error "insufficiently lazy")) <$ s
+  , testProperty "pure and <*> are non-strict" . forall (s :. Nil) $
+    \ s -> void . eval . run $ problematic <$ s
+  ]
+
+strict
+  :: (Traversable f, Monad m)
+  => GenTerm (f ())
+  -> Run f g m
+  -> [TestTree]
+strict s (Run run) =
+  [ testProperty "pure and <*> are strict" . forall (s :. Nil) $
+    \ s -> (liftIO . try @SomeException . evaluate . run $ problematic <$ s) >>= assert . isLeft
```

Comment:
There doesn’t exist a converse to `eval` that ensures its argument throws an exception, so we do it with a combination of `try` and `evaluate`.